src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11PBECipher.java

```diff
@@ -0,0 +1,217 @@
+ */
+
+package sun.security.pkcs11;
+
+import java.security.AlgorithmParameters;
+import java.security.InvalidAlgorithmParameterException;
+import java.security.InvalidKeyException;
+import java.security.Key;
+import java.security.NoSuchAlgorithmException;
+import java.security.SecureRandom;
+import java.security.spec.AlgorithmParameterSpec;
+import java.security.spec.InvalidKeySpecException;
+import javax.crypto.BadPaddingException;
+import javax.crypto.CipherSpi;
+import javax.crypto.IllegalBlockSizeException;
+import javax.crypto.NoSuchPaddingException;
+import javax.crypto.ShortBufferException;
+import javax.crypto.spec.PBEKeySpec;
+import javax.crypto.spec.PBEParameterSpec;
+
+import sun.security.jca.JCAUtil;
+import static sun.security.pkcs11.wrapper.PKCS11Constants.*;
+import sun.security.pkcs11.wrapper.PKCS11Exception;
+import sun.security.util.PBEUtil;
+
+final class P11PBECipher extends CipherSpi {
+    private final Token token;
+    private final String pbeAlg;
+    private final P11Cipher cipher;
+    private final int blkSize;
+    private final P11SecretKeyFactory.PBEKeyInfo svcPbeKi;
+    private final PBEUtil.PBES2Params pbes2Params = new PBEUtil.PBES2Params();
+
+    P11PBECipher(Token token, String pbeAlg, long cipherMech)
+                    throws PKCS11Exception, NoSuchAlgorithmException {
+        super();
+        String cipherTrans;
+        if (cipherMech == CKM_AES_CBC_PAD || cipherMech == CKM_AES_CBC) {
+            cipherTrans = "AES/CBC/PKCS5Padding";
+        } else {
+            throw new NoSuchAlgorithmException(
+                    "Cipher transformation not supported.");
+        }
+        cipher = new P11Cipher(token, cipherTrans, cipherMech);
+        blkSize = cipher.engineGetBlockSize();
+        this.pbeAlg = pbeAlg;
+        svcPbeKi = P11SecretKeyFactory.getPBEKeyInfo(pbeAlg);
+        assert svcPbeKi != null : "algorithm must be in KeyInfo map";
+        this.token = token;
+    }
+
+    // see JCE spec
+    @Override
+    protected void engineSetMode(String mode)
+            throws NoSuchAlgorithmException {
+        cipher.engineSetMode(mode);
+    }
+
+    // see JCE spec
+    @Override
+    protected void engineSetPadding(String padding)
+            throws NoSuchPaddingException {
+        cipher.engineSetPadding(padding);
+    }
+
+    // see JCE spec
+    @Override
+    protected int engineGetBlockSize() {
+        return cipher.engineGetBlockSize();
+    }
+
+    // see JCE spec
+    @Override
+    protected int engineGetOutputSize(int inputLen) {
+        return cipher.engineGetOutputSize(inputLen);
+    }
+
+    // see JCE spec
+    @Override
+    protected byte[] engineGetIV() {
+        return cipher.engineGetIV();
+    }
+
+    // see JCE spec
+    @Override
+    protected AlgorithmParameters engineGetParameters() {
+        return pbes2Params.getAlgorithmParameters(
+                blkSize, pbeAlg, P11Util.getSunJceProvider(),
+                JCAUtil.getSecureRandom());
```

Comment:
The random source should be the one supplied through CipherSpi.engineInit(...) call if there is one available (see line 118). There is Cipher javadoc specifies this.

Reply:
Good point. As I see it, the problem is not in the random source itself but in the values. There are a couple of P11PBECipher::engineInit paths in which P11PBECipher initialization succeeds but the pbes2Params does not have the salt, iCount and ivSpec in use. These paths are those in which the P11 key was already derived (it's a P11PBEKey): we check consistency but record nothing for future P11PBECipher::engineGetParameters calls. I think that we can get the right values from the P11PBEKey and PBEParameterSpec. Notice that if the ivSpec is not passed, it's value could be randomly generated in the underlying Cipher. @franferrax what do you think?

Reply:
We can confirm that the problem was indeed as in my previous comment. As part of fixing this issue, we did some internal refactoring to PBEUtil.PBES2Params which have simplified the code. In particular, PBEUtil.PBES2Params has a method called "initialize" which consolidate all possible state initialization and is called from P11PBECipher::engineGetParameters, PBEUtil.PBES2Params::getPBEKeySpec and PBEUtil.PBES2Params::getAlgorithmParameters.

Reply:
One more comment. You will see JCAUtil.getSecureRandom still passed BUT this is used only if the PBEUtil.PBES2Params instance was not initialized before from engineInit. Notice that if it was initialized before, a random passed to engineInit was used to generate the IV and salt values.